### PR TITLE
MSMR model

### PR DESCRIPTION
**Pull request by rtimms**
# Description

Implements the MSMR model from [Baker & Verbrugge, 2018](https://iopscience.iop.org/article/10.1149/2.0771816jes).

In particular, this PR adds a new "particle" model that solves for the potential within the particles. Then OCP is then this potential evaluated at the surface. In order to use this model, users must write down the OCV for each reaction (in the form of Eq 3), which are inverted to give the fraction of occupied sites (Eq 5). These are summed to give an explicit expression for the inverse of the OCV (Eq 6).

Also updates the electrode State of Health models to work with the MSMR formulation. 

Fixes # (issue)

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all`
- [ ] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

File: pybamm/expression_tree/broadcasts.py
Comment (tinosulzer) [inline pybamm/expression_tree/broadcasts.py:52]: probably cleaner to just get rid of the NotImplementedError for spatial operators

**Review by tinosulzer — COMMENTED** [review]
Looks good, will need some more testing for coverage. Let me know if you want me to review any equations in particular

**Comment by rtimms** [comment]
Thanks. Yep, tests and examples still to do but wanted another pair of eyes on the general implementation.

**Comment by rtimms** [comment]
To do:

- [ ] Refomulate exchange current to use potential and lithiation extent (for numerical stability)
- [ ] Make setting up options cleaner for half cells. At the moment the checks are set up such that you have to pass a tuple of options for the open circuit potential, particle and intercalation kinetics for a half cell even though the values in the li metal side are ignored (or don't make sense in the case of particle)

**Review by brosaplanella — APPROVED** [review]
Looks great, thanks Rob! Just a few comments, happy to merge once tests pass.

File: pybamm/parameters/lithium_ion_parameters.py
Comment (rtimms) [reply pybamm/parameters/lithium_ion_parameters.py:707]: updated the comment to explain this is what the original formulation is in the paper

File: pybamm/models/submodels/particle/base_particle.py
Comment (rtimms) [reply pybamm/models/submodels/particle/base_particle.py:101]: this is a hangover from when we had dimensional and dimensionless concentration. now they are equivalent. maybe we should remove dimensionless concentration now?

File: pybamm/models/full_battery_models/base_battery_model.py
Comment (rtimms) [reply pybamm/models/full_battery_models/base_battery_model.py:265]: thanks, changed to check for positive integers

File: docs/source/examples/notebooks/parameterization/parameterization.ipynb
Comment (rtimms) [reply docs/source/examples/notebooks/parameterization/parameterization.ipynb:1]: this was old and is fixed by re-running

**Pull request merged**
